aiobotocore/credentials.py

```diff
@@ -0,0 +1,774 @@
+    ReadOnlyCredentials, ContainerProvider, ContainerMetadataFetcher, \
+    _parse_if_needed, InstanceMetadataProvider, _get_client_creator, \
+    ProfileProviderBuilder, ConfigProvider, SharedCredentialProvider, \
+    ProcessProvider, AssumeRoleWithWebIdentityProvider, _local_now, \
+    CachedCredentialFetcher, _serialize_if_needed, BaseAssumeRoleCredentialFetcher, \
+    AssumeRoleProvider, AssumeRoleCredentialFetcher, CredentialResolver, \
+    CanonicalNameCredentialSourcer, BotoProvider, OriginalEC2Provider
+from botocore.exceptions import MetadataRetrievalError, CredentialRetrievalError, \
+    InvalidConfigError, PartialCredentialsError, RefreshWithMFAUnsupportedError, \
+    UnknownCredentialError
+from botocore.compat import compat_shell_split
+
+from aiobotocore.utils import AioContainerMetadataFetcher, AioInstanceMetadataFetcher
+from aiobotocore.config import AioConfig
+
+logger = logging.getLogger(__name__)
+
+
+def create_aio_mfa_serial_refresher(actual_refresh):
+    class _Refresher(object):
+        def __init__(self, refresh):
+            self._refresh = refresh
+            self._has_been_called = False
+
+        async def call(self):
+            if self._has_been_called:
+                # We can explore an option in the future to support
+                # reprompting for MFA, but for now we just error out
+                # when the temp creds expire.
+                raise RefreshWithMFAUnsupportedError()
+            self._has_been_called = True
+            return await self._refresh()
+
+    return _Refresher(actual_refresh).call
+
+
+class AioCredentials(Credentials):
+    async def get_frozen_credentials(self):
+        return ReadOnlyCredentials(self.access_key,
+                                   self.secret_key,
+                                   self.token)
+
+    @classmethod
+    def from_credentials(cls, obj: Optional[Credentials]):
+        if obj is None:
+            return None
+        return cls(
+            obj.access_key, obj.secret_key,
+            obj.token, obj.method)
+
+
+class AioRefreshableCredentials(RefreshableCredentials):
+    def __init__(self, *args, **kwargs):
+        super(AioRefreshableCredentials, self).__init__(*args, **kwargs)
+        self._refresh_lock = asyncio.Lock()
+
+        # self._refresh_using should be a coroutine
+
+    @classmethod
+    def from_refreshable_credentials(cls, obj: Optional[RefreshableCredentials]):
+        if obj is None:
+            return None
+        return cls(  # Using interval values here to skip property calling .refresh()
+            obj._access_key, obj._secret_key,
+            obj._token, obj._expiry_time,
+            obj._refresh_using, obj.method,
+            obj._time_fetcher
+        )
+
+    # Redeclaring the properties so it doesnt call refresh
+    # Have to redeclare setter as we're overriding the getter
+    @property
+    def access_key(self):
```

Comment:
@terrycain this is a problem no?  Look at `HmacV1PostAuth.add_auth` : `fields['AWSAccessKeyId'] = self.credentials.access_key`

and `HmacV1QueryAuth._inject_signature` : `query_dict['AWSAccessKeyId'] = self.credentials.access_key`. there are a few more places too

Reply:
this is kinda the lynchpin :)

Reply:
Yeah true

Reply:
Though technically all signing auth classes are passed frozen credentials :D but yeah, the only safe place to read access_key... would be from the ReadOnlyCredential class retrieved from get_frozen_credentials